### PR TITLE
Use commit hash to look in cache instead of calling head

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR tries to limit the calls to requests.head made for cached models every time we try to load them. Currently on the main branch, a call to the following objects results to the following number of underlying calls to the API:
- AutoConfig: 1 (fiou)
- AutoModel: 2 (model + config)
- AutoTokenizer: 9 (multiple tokenizer files and multiple calls to config)
- pipeline: 13 (all of the above + one extra call to config)
- a sharded model: number of shards + 2

This is a bit excessive, so this PR reduces this to the maximum it can by using the commit hash of the first file downloaded: if it's the same as something we have in the cache, then all files in that subfolder with the same commit hash are up to date.

As you can see in the tests it does not completely succeed, because we can't detect with this reasoning if a file does not exist in the repo: if it's not in the cache, it could be because it's still not downloaded yet. But still it reduces the number of calls seen above to:
- AutoConfig: 1
- AutoModel: 1
- AutoTokenizer: between 2 and 4 depending on the tokenizer
- pipeline: between 2 and 4 depending on the tokenizer
- a sharded model: 2

### CODE REVIEW

File: tests/models/auto/test_modeling_tf_auto.py [hunk 1]
```diff
@@ -287,3 +288,21 @@ def test_model_file_not_found(self):
...
+        with RequestCounter() as counter:
+            _ = TFAutoModel.from_pretrained("hf-internal-testing/tiny-random-bert")
+            self.assertEqual(counter.get_request_count, 0)
+            self.assertEqual(counter.head_request_count, 1)
+            self.assertEqual(counter.other_request_count, 0)
+
+        # With a sharded checkpoint
+        _ = TFAutoModel.from_pretrained("ArthurZ/tiny-random-bert-sharded")
+        with RequestCounter() as counter:
+            _ = TFAutoModel.from_pretrained("ArthurZ/tiny-random-bert-sharded")
```
Comment (LysandreJik) [inline tests/models/auto/test_modeling_tf_auto.py:304]: Could this re-use a model from `hf-internal-testing`?

**Review by LysandreJik — APPROVED** [review]
This is an incredible improvement! Thanks for going the extra mile and adding some tests.

Let's have this sit on `main` for a while before releasing to ensure that it doesn't break anything, but it looks great :ok_hand:

**Review by Narsil — APPROVED** [review]
LGTM

File: src/transformers/utils/hub.py [hunk 2]
```diff
@@ -816,6 +848,7 @@ def get_checkpoint_shard_files(
     """
     import json
 
+    print(_commit_hash)
```
Comment (julien-c) [inline src/transformers/utils/hub.py:851]:
```suggestion
```

**Review by Wauplin — APPROVED** [review]

**Review by SBrandeis — APPROVED** [review]

File: src/transformers/utils/hub.py [hunk 3]
```diff
@@ -200,11 +200,26 @@ def http_user_agent(user_agent: Union[Dict, str, None] = None) -> str:
     return ua
 
 
-def try_to_load_from_cache(cache_dir, repo_id, filename, revision=None):
+def extract_commit_hash(resolved_file, commit_hash):
```
Comment (julien-c) [inline src/transformers/utils/hub.py:203]: (nit) personally I'm really lost without type hints :)

File: src/transformers/utils/hub.py [hunk 4]
```diff
@@ -200,11 +200,26 @@ def http_user_agent(user_agent: Union[Dict, str, None] = None) -> str:
...
+    """
+    Extracts the commit hash from a resolved filename toward a cache file.
+    """
+    if resolved_file is None or commit_hash is not None:
+        return commit_hash
+
+    search = re.search(r"snapshots/([^/]+)/", resolved_file)
```
Comment (julien-c) [inline src/transformers/utils/hub.py:210]: potentially robustify this by looking for a hex string of length commit_hash_length

File: src/transformers/testing_utils.py [hunk 5]
```diff
@@ -1588,3 +1589,30 @@ def run_command(command: List[str], return_stdout=False):
...
+
+    def new_request(self, method, **kwargs):
+        if method == "GET":
+            self.get_request_count += 1
+        elif method == "HEAD":
+            self.head_request_count += 1
+        else:
+            self.other_request_count += 1
+
+        return self.old_request(method=method, **kwargs)
```
Comment (julien-c) [inline src/transformers/testing_utils.py:1618]: OMG Python 🤯

**Review by julien-c — APPROVED** [review]
my only nit is that we shouldn't really need `extract_commit_hash` because `cached_file` knows this info so could return it somehow, and parsing a filepath looks a bit unclean to me. (But i can understand limiting methods' API changes to a minimum)

I didn't follow the full flow of `_commit_hash` being passed around but it looks good to me from a quick glance

File: src/transformers/utils/hub.py (hunk 4 above)
Comment (julien-c) [reply src/transformers/utils/hub.py:210]: (BTW could also potentially live in `huggingface_hub`)

File: src/transformers/testing_utils.py (hunk 5 above)
Comment (sgugger) [reply src/transformers/testing_utils.py:1618]: Black Python magic ;-) 🪄

File: src/transformers/utils/hub.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/utils/hub.py:210]: Reused REGEX_COMMIT_HASH from huggingface_hub here.

**Pull request merged**